File: 0200-number-of-islands/0200-number-of-islands.py
from collections import deque

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        row_len, column_len = len(grid), len(grid[0])
        direction = [[0,1], [-1,0], [0,-1], [1,0]]
        visited = [[False] * column_len for _ in range(row_len)]

        def bfs(grid, row, column):
            nonlocal count, visited

            q = deque()
            q.append((row, column))
            visited[row][column] = True

            while q:
                current_r, current_c = q.popleft()

                for dr, dc in direction:
                    next_r = current_r + dr
                    next_c = current_c + dc

                    if 0 <= next_r < row_len and 0 <= next_c < column_len and grid[next_r][next_c] == "1":
                        if not visited[next_r][next_c]:
                            q.append((next_r, next_c))
                            visited[next_r][next_c] = True

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if(grid[i][j] == "1" and not visited[i][j]):
                    count += 1
                    bfs(grid, i, j)
        
        return count
        

File: LeetCode/Medium/0200-number-of-islands/Java/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];

        int count = 0;

        for(int i=0; i<grid.length; i++) {
            for(int j=0; j<grid[0].length; j++) {
                if(!visited[i][j] && grid[i][j] == '1') {
                    count += 1;
                    dfs(i, j, visited, grid);
                }
            }
        }

        return count;
    }

    private void dfs(int x, int y, boolean[][] visited, char[][] grid) {
        int[][] direction = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
        visited[x][y] = true;

        for(int d[] : direction) {
            int next_x = x + d[0];
            int next_y = y + d[1];

            if(next_x >= 0 && next_y >= 0 && next_x < visited.length && next_y < visited[0].length) {
                if(!visited[next_x][next_y] && grid[next_x][next_y] == '1') {
                    dfs(next_x, next_y, visited, grid);
                }
            }
        }
    }
}